tools/taish/tai_shell.cpp

```diff
@@ -0,0 +1,593 @@
+
+#include <dlfcn.h>
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/ip.h> 
+#include <ext/stdio_filebuf.h> 
+#include <poll.h> 
+#include <arpa/inet.h>
+
+#include "tai.h"
+#include "tai_shell.hpp"
+
+tai_api *p_tai_api;
+
+std::map<std::string, tai_command_fn> tai_cli_shell::cmd2handler = {
+   {"?", tai_command_help},
+   {"help", tai_command_help},
+   {"load", tai_command_load},
+   {"init", tai_command_init},
+   {"quit", tai_command_quit},
+   {"exit", tai_command_quit},
+   {"logset", tai_command_logset}
+};
+
+std::vector <std::string> help_msgs = {
+  {"?: show help messages for all commands\n"},
+  {"help: show help messages for all commands\n"},
+  {"load: load a TAI librarary: Usage: load <TAI librabry file name> \n"},
+  {"init: Initialize TAI API.: Usage: init\n"},
+  {"quit: Quit this telnet session.\n"},
+  {"exit: Exit this telnet session.\n"},
+  {"logset: Set log level.: Usage: logset <debug|info|notice|warn:error|critical> \n"},
+};
+
+tai_module_api_t *module_api;
+tai_network_interface_api_t *netif_api;
+tai_host_interface_api_t *hostif_api;
+
+int fd;
+std::queue<std::pair<bool, std::string>> q;
+std::mutex m;
+
+class module {
+    public:
+        module(tai_object_id_t id) : m_id(id) {
+            std::vector<tai_attribute_t> list;
+            tai_attribute_t attr;
+            attr.id = TAI_MODULE_ATTR_NUM_HOST_INTERFACES;
+            list.push_back(attr);
+            attr.id = TAI_MODULE_ATTR_NUM_NETWORK_INTERFACES;
+            list.push_back(attr);
+            auto status = module_api->get_module_attributes(id, list.size(), list.data());
+            if ( status != TAI_STATUS_SUCCESS ) {
+                throw std::runtime_error("faile to get attribute");
+            }
+            std::cout << "num hostif: " << list[0].value.u32 << std::endl;
+            std::cout << "num netif: " << list[1].value.u32 << std::endl;
+            create_hostif(list[0].value.u32);
+            create_netif(list[1].value.u32);
+        }
+    private:
+        tai_object_id_t m_id;
+        std::vector<tai_object_id_t> netifs;
+        std::vector<tai_object_id_t> hostifs;
+        int create_hostif(uint32_t num);
+        int create_netif(uint32_t num);
+        int loop();
+};
+
+int module::create_netif(uint32_t num) {
+    for ( int i = 0; i < num; i++ ) {
+        tai_object_id_t id;
+        std::vector<tai_attribute_t> list;
+        tai_attribute_t attr;
+        attr.id = TAI_NETWORK_INTERFACE_ATTR_INDEX;
+        attr.value.u32 = i;
+        list.push_back(attr);
+        auto status = netif_api->create_network_interface(&id, m_id, list.size(), list.data());
+        if ( status != TAI_STATUS_SUCCESS ) {
+            throw std::runtime_error("failed to create network interface");
+        }
+        std::cout << "netif: " << id << std::endl;
+        netifs.push_back(id);
+
+        list.clear();
+
+        attr.id = TAI_NETWORK_INTERFACE_ATTR_TX_ENABLE;
```

Comment:
As per the discussion [here](https://github.com/Telecominfraproject/oopt-tai/pull/25#discussion_r216023892), please remove lines setting non-mandatory attributes.